// @dendronhq/common-all/lib/abTests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CURRENT_AB_TESTS = exports.AB_TUTORIAL_TEST = exports.TutorialTestGroups = exports.GRAPH_THEME_FEATURE_SHOWCASE_TEST = exports.GraphThemeFeatureShowcaseTestGroups = exports.GRAPH_THEME_TEST = exports.GraphThemeTestGroups = exports.MeetingNoteTestGroups = void 0;
// ^xi5t1r2j51ot
const abTesting_1 = require("./abTesting");
/**
 * Section: Tests (Active or soon to be active)
 *
 * NOTE: please follow this convention for naming future tests:
 * YYYY-MM-TEST_NAME.  For example, 2022-04-MEETING_NOTE_FEATURE_SHOWCASE.
 *
 * See [[A/B Testing|dendron://dendron.docs/ref.ab-testing]] for more details.
 */
var MeetingNoteTestGroups;
(function (MeetingNoteTestGroups) {
    MeetingNoteTestGroups["show"] = "show";
    MeetingNoteTestGroups["noShow"] = "noShow";
})(MeetingNoteTestGroups = exports.MeetingNoteTestGroups || (exports.MeetingNoteTestGroups = {}));
var GraphThemeTestGroups;
(function (GraphThemeTestGroups) {
    /**
     * New user will get Monokai graph theme by default
     */
    GraphThemeTestGroups["monokai"] = "Monokai";
    /**
     * New user will get Classic graph theme by default
     */
    GraphThemeTestGroups["classic"] = "Classic";
    /**
     * New User will get Block theme by default
     */
    GraphThemeTestGroups["block"] = "Block";
})(GraphThemeTestGroups = exports.GraphThemeTestGroups || (exports.GraphThemeTestGroups = {}));
exports.GRAPH_THEME_TEST = new abTesting_1.ABTest("GraphThemeTest", [
    {
        name: GraphThemeTestGroups.monokai,
        weight: 1,
    },
    {
        name: GraphThemeTestGroups.classic,
        weight: 1,
    },
    {
        name: GraphThemeTestGroups.block,
        weight: 1,
    },
]);
var GraphThemeFeatureShowcaseTestGroups;
(function (GraphThemeFeatureShowcaseTestGroups) {
    GraphThemeFeatureShowcaseTestGroups["showMeHow"] = "showMeHow";
    GraphThemeFeatureShowcaseTestGroups["openGraph"] = "openGraphView";
})(GraphThemeFeatureShowcaseTestGroups = exports.GraphThemeFeatureShowcaseTestGroups || (exports.GraphThemeFeatureShowcaseTestGroups = {}));
exports.GRAPH_THEME_FEATURE_SHOWCASE_TEST = new abTesting_1.ABTest("GraphThemeFeatureShowcaseTest", [
    {
        name: GraphThemeFeatureShowcaseTestGroups.openGraph,
        weight: 1,
    },
    {
        name: GraphThemeFeatureShowcaseTestGroups.showMeHow,
        weight: 1,
    },
]);
/**
 *
 * These enums correspond to the alt treatment hierarchy, and will have to match
 * in order to be synced correctly / loaded by the tutorial initializer.
 *
 * e.g.)
 *  1. in `dendron-site`, prepare tutorials in `tutorial.main.*`, `tutorial.my-treatment-1.*`, and `tutorial.my-treatment-2.*`,
 *  2. in this case, the enum kvps should match the second level hierarchy:
 *     export enum TutorialTestGroups {
 *       "main" = "main",
 *       "my-treatment-1" = "my-treatment-1",
 *       "my-treatment-2" = "my-treatment-2"
 *     }
 */
var TutorialTestGroups;
(function (TutorialTestGroups) {
    TutorialTestGroups["main"] = "main";
    TutorialTestGroups["no-refactor"] = "no-refactor";
    TutorialTestGroups["original"] = "original";
})(TutorialTestGroups = exports.TutorialTestGroups || (exports.TutorialTestGroups = {}));
/**
 * Long running A/B test for differently treated tutorials.
 * See instructions for how to set up treatments in the comments for {@link TutorialTestGroups}
 *
 * Currently running test:
 *   - old tutorial (including meeting note mention) as baseline
 *   - one group that will get the updated tutorial that is presented in our wiki
 *   - one group that will get the updated tutorial that is presented in our wiki, minus the mentions to refactoring.
 */
exports.AB_TUTORIAL_TEST = new abTesting_1.ABTest("AB_TUTORIAL_TEST", [
    {
        name: TutorialTestGroups["main"],
        weight: 1,
    },
    {
        name: TutorialTestGroups["no-refactor"],
        weight: 1,
    },
    {
        name: TutorialTestGroups["original"],
        weight: 1,
    },
]);
/** All A/B tests that are currently running.
 *
 * ^tkqhy45hflfd
 */
exports.CURRENT_AB_TESTS = [
    exports.GRAPH_THEME_TEST,
    exports.GRAPH_THEME_FEATURE_SHOWCASE_TEST,
    exports.AB_TUTORIAL_TEST,
];
